piqueserver/game_modes/push.py

```diff
@@ -69,131 +109,105 @@ def byte_middle_range(byte):
+
+
+@command('r')
+def respawn(connection):
+    if not ALLOW_RESPAWN_COMMAND:
+        return
+    if connection.world_object is not None and not connection.world_object.dead:
+        if (connection.last_spawn_time is None or connection.last_spawn_time +
+                RESPAWN_CMD_DELAY <= get_now_in_secs()):
+            if has_flag(connection):
+                connection.drop_flag()
+            connection.spawn()
+        else:
+            connection.send_chat(
+                "Please wait %s seconds before using this command again."
+                % (connection.last_spawn_time + RESPAWN_CMD_DELAY - get_now_in_secs()))
+
+
+def get_entity_location(connection, entity_id):
     if entity_id == BLUE_BASE:
-        return self.protocol.blue_team.cp
+        return connection.protocol.blue_team.cp
     elif entity_id == GREEN_BASE:
-        return self.protocol.green_team.cp
+        return connection.protocol.green_team.cp
 
-    # this next part might seem counter intuitive but you need the blue intel
-    # to spawn near the greens and vice versa
     elif entity_id == BLUE_FLAG:
-        return self.protocol.green_team.spawn
+        return (connection.protocol.green_team.spawn[0],
+                connection.protocol.green_team.spawn[1],
+                1)
     elif entity_id == GREEN_FLAG:
-        return self.protocol.blue_team.spawn
+        return (connection.protocol.blue_team.spawn[0],
+                connection.protocol.blue_team.spawn[1],
+                1)
 
 
 def get_spawn_location(connection):
     # distance from spawn center to randomly spawn in
     spawn_range = connection.protocol.spawn_range
-
-    if connection.team.spawn is not None:
-        xb = connection.team.spawn[0]
-        yb = connection.team.spawn[1]
-        xb += randint(-spawn_range, spawn_range)
-        yb += randint(-spawn_range, spawn_range)
-        return (xb, yb, connection.protocol.map.get_z(xb, yb))
+    xb = connection.team.spawn[0]
+    yb = connection.team.spawn[1]
+    xb += randint(-spawn_range, spawn_range)
+    yb += randint(-spawn_range, spawn_range)
+    return (xb, yb, connection.protocol.map.get_z(xb, yb))
 
 
 def apply_script(protocol, connection, config):
     class PushConnection(connection):
+        last_spawn_time = None
+        # list entry format: ((x, y, z), timestamp when block was placed)
+        last_blocks = None
 
-        def invalid_build_position_check(
-                self, x, y, check_area, error_message):
-            # check_area is made of x-pos, y-pos of upper left corner,
-            # and x-pos, y-pos of opposite corner (bottom right)
-            if (x >= check_area[0] and y >= check_area[1] and
-                    x <= check_area[2] and y <= check_area[3]):
+        def is_in_invalid_area(self, x, y, check_area, error_message):
+            if is_in_area(x, y, check_area[0], check_area[1],
+                          check_area[2], check_area[3]):
```

Comment:
replace with `is_in_are(x, y, *check_area)`. This auto expands a sequence into the arguments.